File: Middle-logic/Models/ObservableObject.js
import { SensoryStimulus } from "./SensoryStimulus.js";
import { MentalObject } from "./MentalObject.js";

export class ObservableObjectType{
    static SensoryStimuls = "Sensory Stimulus";
    static MentalObject = "Mental Object";
}
export class ObservableObjectIntensity{
    static Mild = "Mild";
    static Moderate = "Moderate";
    static Intense = "Intense";
}
export class ObservableObject{
    constructor(id, title, type, description, icon, intensity){
        this.Observable_Object_ID = id;
        this.Title = title;
        this.Type = type;
        this.Description = description;
        this.Icon = icon;
        this.Intensity = intensity;
    }
    static getObservableObjectFromObject(obj){
        var observableObject;
        if (obj.Type == ObservableObjectType.MentalObject) {
            observableObject = new MentalObject();
            observableObject.Mental_Object_Type = obj.Mental_Object_Type;
        } else {
            observableObject = new SensoryStimulus();
            observableObject.Sensory_Stimulus_Type = obj.Sensory_Stimulus_Type;
        }

        observableObject.Observable_Object_ID = obj.Observable_Object_ID;
        observableObject.Title = obj.Title;
        observableObject.Type = obj.Type;
        observableObject.Description = obj.Description;
        observableObject.Icon = obj.Icon;
        observableObject.Intensity = obj.Intensity;

        return observableObject;
    }
}

File: backend/Models/ObservableObject.cs
using static backend.Models.ObservableObject;

namespace backend.Models
{
    public enum ObservableObjectType
    {
        MentalObject,
        SensoryStimulus
    }
    public enum ObservableObjectSubType
    {
        Visual,
        Auditory,
        Olfactory,
        Kinesthetic,
        Taste,

        Thought,
        MentalState,
        FeelingTone
    }
    public enum IntensityType
    {
        Mild,
        Moderate,
        Intense
    }
    public enum Proximity
    {
        Unrelated,
        SameSubType,
        DirectlyRelated,
        MeditationObject
    }
    public enum FeelingTone
    {
        Pleasant,
        Unpleasant,
        Neutral
    }
    public class ObservableObject
    {
        public int Id { get; set; }
        public string Title { get; set; } = String.Empty;
        public string? Description { get; set; }
        public IntensityType? Intensity { get; set; }
        public ObservableObjectSubType SubType { get; set; }
        public Proximity ProximityToMO { get; set; }

        public ObservableObjectType Type()
        {
            if (
                SubType == ObservableObjectSubType.Visual ||
                SubType == ObservableObjectSubType.Auditory ||
                SubType == ObservableObjectSubType.Olfactory ||
                SubType == ObservableObjectSubType.Kinesthetic ||
                SubType == ObservableObjectSubType.Taste
            )
            {
                return ObservableObjectType.SensoryStimulus;
            }
            else
            {
                return ObservableObjectType.MentalObject;
            }
        }
        public ObservableObject()
        {

        }
        public ObservableObject(string title, string description, string Icon,
        IntensityType intensity, ObservableObjectType type, ObservableObjectSubType subType,
        FeelingTone tone)
        {
            this.Title = title;
            this.Description = description;
            this.Intensity = intensity;
            this.SubType = subType;
        }
    }
}
